test/xbgpu/test_engine.py

```diff
@@ -336,25 +336,36 @@ async def _send_data(
         max_heaps = n_ants * HEAPS_PER_FENGINE_PER_CHUNK * 10
         feng_stream = self._make_feng(mock_recv_streams, max_packet_size, max_heaps)
 
-        acc_index_counts: list[dict[int, int]] = []
+        # NOTE: The data generation creates an extra heap or several for the
+        # XPipeline with the smallest `heap_accumulation_threshold`. This results
+        # in an incomplete final accumulation for that XPipeline. Incomplete
+        # accumulations are purposefully dictated by `batch_indices` constructed
+        # by the caller.
+        # For these reasons, `acc_index_heap_counts` tracks the heap count for
+        # each accumulation index (for each corrprod output). If a heap count
+        # for a corrprod's final accumulation is not equal to its
+        # `heap_accumulation_threshold`, the accumulation index is discarded
+        # from the list (used later on). If this function attempted to receive
+        # data for an XPipeline's final, incomplete accumulation, it would
+        # result in a _spead2.Stopped error as that accumulation would never
+        # have been processed/transmitted by the XPipeline.
+        # - i.e. the test would attempt to get one more heap than the stream
+        #   contained.
+        acc_index_heap_counts: list[dict[int, int]] = []
```

Comment:
Is the whole exercise of determining the LCM of the two heap accumulation thresholds not to ensure that this doesn't happen? Or did I miss something?

Reply:
I'd added in this logic to resolve the `spead2 ring buffer has been stopped` error, before we delved further into the 1A debug. I can look into it further now if you want (to see if it's still necessary), but I'm leaning towards making this part of the eventual technical-debt ticket for test cleanup.

Reply:
Yeah IMO just leave it for now. It currently works. We will probably have to make several tickets but that's OK.